File: Stats.CPP/EWMA_Demo.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "EWMA_Demo.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "Basic_Form"
#pragma resource "*.dfm"
TfrmEWMA *frmEWMA;
//---------------------------------------------------------------------------
__fastcall TfrmEWMA::TfrmEWMA(TComponent* Owner)
	: TfrmBasic(Owner)
{
	Memo1->Lines->Add("New in release 1.1: support for Exponential weighted moving average chart (EWMA)");
	Memo1->Lines->Add(" This example shows you how to construct EWMA chart. EWMAChart procedure supports "
 		"asymptote value and exact value for control limits.");

	Data->LoadFromFile(CodePath() + "ewma_data.mtx");

	r = 0.35;
	Confidence = 0.9987;
	Edit1->Text = FloatToStr(r);
	Edit2->Text = FloatToStr(Confidence);
	RadioGroup1Click(RadioGroup1);
}
//---------------------------------------------------------------------------
void __fastcall TfrmEWMA::RadioGroup1Click(TObject *Sender)
{
	Chart1->FreeAllSeries();
	switch (RadioGroup1->ItemIndex)
	{
		case 0:  {
					Chart1->AddSeries(new TQCSeries(Chart1));
					EWMAChart(Data,DrawVec,cl,ucl,lcl,r,Confidence);
					Chart1->Legend->Visible = false;
					DrawValues(DrawVec,Chart1->Series[0],0,1,false);
					Chart1->Series[0]->SeriesColor = clBlue;
					((TQCSeries*) Chart1->Series[0])->ControlLimitPen->Color = clRed;
					((TQCSeries*) Chart1->Series[0])->CL = cl;
					((TQCSeries*) Chart1->Series[0])->UCL = ucl;
					((TQCSeries*) Chart1->Series[0])->LCL = lcl;
				};break;
		case 1:  {
					Chart1->AddSeries(new TLineSeries(Chart1));
					Chart1->AddSeries(new TLineSeries(Chart1));
					Chart1->AddSeries(new TLineSeries(Chart1));
					EWMAChart(Data,DrawVec,cl,vucl,vlcl,r,Confidence);
					Chart1->Legend->Visible = true;
					((TLineSeries*)Chart1->Series[0])->Pointer->Visible = true;
					((TLineSeries*)Chart1->Series[0])->SeriesColor = clBlue;
					((TLineSeries*)Chart1->Series[0])->Title = "Data";
					Chart1->Series[1]->SeriesColor = clRed;
					((TLineSeries*)Chart1->Series[1])->Pen->Style = psDot;
					Chart1->Series[1]->Title = "UCL";
					Chart1->Series[2]->SeriesColor = clRed;
					((TLineSeries*)Chart1->Series[2])->Pen->Style = psDot;
					Chart1->Series[2]->Title = "LCL";
					DrawValues(DrawVec,Chart1->Series[0],0,1,false);
					DrawValues(vucl,Chart1->Series[1],0,1,false);
					DrawValues(vlcl,Chart1->Series[2],0,1,false);
				};break;
	}
}
//---------------------------------------------------------------------------
void __fastcall TfrmEWMA::Button1Click(TObject *Sender)
{
	ViewValues(Data,"EWMA data",true,false);
}
//---------------------------------------------------------------------------
void __fastcall TfrmEWMA::Edit1Change(TObject *Sender)
{
	if (Showing)
	{
		try
		{
			r = StrToFloat(Edit1->Text);
		}
		catch (...) {
			r = 0.35;
		}
		RadioGroup1Click(RadioGroup1);
	}
}
//---------------------------------------------------------------------------
void __fastcall TfrmEWMA::Edit2Change(TObject *Sender)
{
	if (Showing)
	{
		try
		{
			Confidence = StrToFloat(Edit2->Text);
		}
		catch (...) {
			Confidence = 0.9987;
		}
		RadioGroup1Click(RadioGroup1);
	}
}
//---------------------------------------------------------------------------
void __fastcall TfrmEWMA::Chart1AfterDraw(TObject *Sender)
{
	if (RadioGroup1->ItemIndex == 1) {
		Chart1->Canvas->Pen->Color = clBlue;
		Chart1->Canvas->MoveTo(Chart1->ChartRect.Left,Chart1->LeftAxis->CalcPosValue(cl));
		Chart1->Canvas->ClipRectangle(Chart1->ChartRect);
		Chart1->Canvas->LineTo(Chart1->ChartRect.Right,Chart1->LeftAxis->CalcPosValue(cl));
		Chart1->Canvas->UnClipRectangle();
	}
}
//---------------------------------------------------------------------------

File: Stats.Delphi/EWMA_Demo.pas
unit EWMA_Demo;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  Basic_Form, TeeProcs, TeEngine, Chart, StdCtrls, ExtCtrls,
  MtxVec, Math387;

type
  TfrmEWMA = class(TfrmBasic)
    Panel2: TPanel;
    Chart1: TChart;
    RadioGroup1: TRadioGroup;
    Button1: TButton;
    Label1: TLabel;
    Label2: TLabel;
    Edit1: TEdit;
    Edit2: TEdit;
    procedure RadioGroup1Click(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure Button1Click(Sender: TObject);
    procedure Edit1Change(Sender: TObject);
    procedure Edit2Change(Sender: TObject);
    procedure Chart1AfterDraw(Sender: TObject);
  private
    { Private declarations }
    Data: TMtx;
    vucl,vlcl, DrawVec: TVec;                          
    ucl,lcl,cl, r, Confidence: double;
  public
    { Public declarations }
  end;

var
  frmEWMA: TfrmEWMA;

implementation

Uses Series, StatSeries, MtxVecTee, MtxVecEdit, StatControlCharts;

{$R *.DFM}

procedure TfrmEWMA.RadioGroup1Click(Sender: TObject);
begin
  Chart1.FreeAllSeries;
  case RadioGroup1.ItemIndex of
    0:  begin
          Chart1.AddSeries(TQCSeries.Create(Chart1));
          EWMAChart(Data,DrawVec,cl,ucl,lcl,r,Confidence);
          Chart1.Legend.Visible := False;
          DrawValues(DrawVec,Chart1.Series[0],0,1,false);
          Chart1.Series[0].SeriesColor := clBlue;
          (Chart1.Series[0] as TQCSeries).ControlLimitPen.Color := clRed;
          (Chart1.Series[0] as TQCSeries).CL := cl;
          (Chart1.Series[0] as TQCSeries).UCL := ucl;
          (Chart1.Series[0] as TQCSeries).LCL := lcl;
        end;
    1:  begin
          Chart1.AddSeries(TLineSeries.Create(Chart1));
          Chart1.AddSeries(TLineSeries.Create(Chart1));
          Chart1.AddSeries(TLineSeries.Create(Chart1));
          EWMAChart(Data,DrawVec,cl,vucl,vlcl,r,Confidence);
          Chart1.Legend.Visible := True;
          With Chart1.Series[0] as TLineSeries do
          begin
            Pointer.Visible := True;
            SeriesColor := clBlue;
            Title := 'Data';
          end;
          Chart1.Series[1].SeriesColor := clRed;
          (Chart1.Series[1] as TLineSeries).Pen.Style := psDot;
          Chart1.Series[1].Title := 'UCL';
          Chart1.Series[2].SeriesColor := clRed;
          (Chart1.Series[2] as TLineSeries).Pen.Style := psDot;
          Chart1.Series[2].Title := 'LCL';
          DrawValues(DrawVec,Chart1.Series[0],0,1,false);
          DrawValues(vucl,Chart1.Series[1],0,1,false);
          DrawValues(vlcl,Chart1.Series[2],0,1,false);
        end;
    end;
end;

procedure TfrmEWMA.FormCreate(Sender: TObject);
begin
  inherited;
  With Memo1.Lines do
  begin
    Clear;
    Add('New in release 1.1: support for Exponential weighted moving average chart (EWMA)');
    Add(' This example shows you how to construct EWMA chart. EWMAChart procedure supports '+
        'asymptote value and exact value for control limits.');
  end;

  Data := TMtx.Create;
  DrawVec := TVec.Create;
  vucl := TVec.Create;
  vlcl := TVec.Create;
  try
    Data.LoadFromFile('ewma_data.mtx');
  except
  end;
  r := 0.35;
  Confidence := 0.9987;
  Edit1.Text := FloatToStr(r);
  Edit2.Text := FloatToStr(Confidence);
  RadioGroup1Click(RadioGroup1);
end;

procedure TfrmEWMA.FormDestroy(Sender: TObject);
begin
  Data.Free;
  DrawVec.Free;
  vucl.Free;
  vlcl.Free;
  inherited;
end;

procedure TfrmEWMA.Button1Click(Sender: TObject);
begin
  ViewValues(Data,'EWMA data',true,false);
end;

procedure TfrmEWMA.Edit1Change(Sender: TObject);
begin
  if Showing then
  begin
    try
      r := StrToFloat(Edit1.Text);
    except
      r := 0.35;
    end;
    RadioGroup1Click(RadioGroup1);
  end;
end;

procedure TfrmEWMA.Edit2Change(Sender: TObject);
begin
  if Showing then
  begin
    try
      Confidence := StrToFloat(Edit2.Text);
    except
      Confidence := 0.9987;
    end;
    RadioGroup1Click(RadioGroup1);
  end;
end;

procedure TfrmEWMA.Chart1AfterDraw(Sender: TObject);
begin
  if RadioGroup1.ItemIndex = 1 then
  with Chart1.Canvas, Chart1 do
  begin
    Pen.Color := clBlue;
    MoveTo(ChartRect.Left,Chart1.LeftAxis.CalcPosValue(cl));
    ClipRectangle(ChartRect);
    LineTo(ChartRect.Right,Chart1.LeftAxis.CalcPosValue(cl));
    UnclipRectangle;
  end;
end;

Initialization
  RegisterClass(TfrmEWMA);

end.

